core/include/traccc/edm/track_state.hpp

```diff
@@ -71,20 +71,34 @@ struct track_state {
 
     /// @return the local position of measurement with 2 X 1 matrix
     // FIXME: The conversion from vector to matrix is inefficient
-    TRACCC_HOST_DEVICE
-    inline matrix_type<2, 1> measurement_local() const {
-        matrix_type<2, 1> ret = matrix_operator().template zero<2, 1>();
-        matrix_operator().element(ret, 0, 0) = m_measurement.local[0];
-        matrix_operator().element(ret, 1, 0) = m_measurement.local[1];
+    template <size_type D>
+    TRACCC_HOST_DEVICE inline matrix_type<D, 1> measurement_local() const {
+        static_assert(((D == 1u) || (D == 2u)),
+                      "The measurement dimension should be 1 or 2");
+
+        matrix_type<D, 1> ret = matrix_operator().template zero<D, 1>();
+        if constexpr (D == 1u) {
+            matrix_operator().element(ret, 0, 0) = m_measurement.local[0];
+        } else if (D == 2u) {
+            matrix_operator().element(ret, 0, 0) = m_measurement.local[0];
+            matrix_operator().element(ret, 1, 0) = m_measurement.local[1];
+        }
```

Comment:
Also, the generation of the zero matrix is probably not needed here?

Reply:
Isn't ```matrix_type<D, 1>``` just a vector and you are setting all of its elements explicitly anyway?

Reply:
I need to set them explicitly. otherwise it can have undefined values.
matrix_type<D ,1> is a vector but I need it to be matrix type to use it in matrix operations.

Reply:
My point is, aren't you setting them explicitly twice now? First to {0, 0} and then to {loc[0], loc[1]} ? Anyway, it's a minor point, so let's just leave it as it is...

Reply:
Understand. Fixed it now